### PR TITLE
 [hlx up] Requesting / fires an error

### PR DESCRIPTION
**Pull request by kptdobe**
PR for https://github.com/adobe/petridish/issues/70

### CODE REVIEW

**Comment by kptdobe** [comment]
Fallback to README is more complicated. For now, only mapping properly / to /index.html

**Comment by trieloff** [comment]
Shouldn't we read this from the `index` configuration property in the strain?

**Comment by kptdobe** [comment]
Pb is that Petridish does not know about strains (yet)... It does not read the strains.yaml file nor having something strain related.

**Comment by kptdobe** [comment]
I improved the code to use the computed index (index.md or README.md): computation is made by HelixServer on init. Better than before and cover the 2 files.

**Comment by kptdobe** [comment]
I also change the `checkPaths()` code which I think was wrong: priority was given to README.md (computed after index.md) - if both are here, README.md wins which should be the other way around.

**Comment by kptdobe** [comment]
Need to fix the tests.

File: src/HelixProject.js [hunk 1]
```diff
@@ -163,6 +163,14 @@ class HelixProject {
     return this._server;
   }
 
+  /**
+   * Returns the index resource path: /index or /README
+   */
+  get index() {
```
Comment (tripodsan) [inline src/HelixProject.js]: why no just return `this._index` ?

File: src/HelixProject.js [hunk 2]
```diff
@@ -171,16 +179,16 @@ class HelixProject {
   }
 
   async checkPaths() {
-    const idxPath = path.resolve(this._cwd, INDEX_MD);
```
Comment (tripodsan) [inline src/HelixProject.js]: I would just guard it with an if. so you can set it from external

File: src/HelixProject.js [hunk 3]
```diff
@@ -171,16 +179,16 @@ class HelixProject {
...
-    if (await isFile(idxPath)) {
-      this._indexMd = idxPath;
-    }
-
     const readmePath = path.join(this._cwd, README_MD);
```
Comment (tripodsan) [inline src/HelixProject.js]: I would just guard it with an if. so you can set it from external

File: src/RequestContext.js [hunk 4]
```diff
@@ -26,6 +26,11 @@ module.exports = class RequestContext {
     this._method = req.method || 'GET';
     this._params = req.query || {};
 
+    if (this._path === '/') {
```
Comment (tripodsan) [inline src/RequestContext.js]: needs to work for all directories, not just `/`

Comment (kptdobe) [reply src/RequestContext.js]: Actually, that's not so simple. How do you know if `/test` must be rendered as `/test` or as `/test/index.html`. For now the only possible distinction I see is the extension, the pipeline is related to an extension. If no extension, then it is a directory. But that's good only until we introduce the no extension pipeline or some "mount" path.

File: src/HelixProject.js (hunk 2 above)
Comment (kptdobe) [reply src/HelixProject.js]: I do understand what you mean. I did not change that code (and do not want to), I just swapped the order of execution.

File: src/HelixProject.js (hunk 1 above)
Comment (kptdobe) [reply src/HelixProject.js]: Linting complains with `this._cfg._index` (2x underscores). And it is nicer to have a method. 
If the comment is about the content of the method, then yes I can compute a `_index` and just return it.

**Comment by kptdobe** [comment]
The fact is that we need to decide based on the path requested without knowing the actually folder / file structure. As mentioned above, the only rule I see for now is if the requested path does not have an extension, then automatically append the index.
And yes, we could have an extra config somewhere to define what is the default index (I like though the fact that we compute it automatically for local content)

**Comment by trieloff** [comment]
> As mentioned above, the only rule I see for now is if the requested path does not have an extension, then automatically append the index.

That would be consistent with our VCL.

File: src/HelixProject.js (hunk 1 above)
Comment (tripodsan) [reply src/HelixProject.js]: :-)

no, I mean, why do you take the substring of indexMd?

**Comment by tripodsan** [comment]
I suggest:

1. remove `this._indexMd` and all that logic
2. add simple `this._directoryIndex`, `withDirectoryIndex()`, `get directoryIndex()`
3. if path has no `.` after last slash, append `this._directoryIndex`

File: src/RequestContext.js (hunk 4 above)
Comment (tripodsan) [reply src/RequestContext.js]: we can ignore this problem for now. if path has no `.` after last slash, append this._index

**Comment by kptdobe** [comment]
I added: 
2. add simple this._directoryIndex, withDirectoryIndex(), get directoryIndex()
3. if path has no . after last slash, append this._directoryIndex

I wonder if this is worth for me spending more time on this since every will change again with #76, especially the `this._indexMd and all that logic`.

File: src/RequestContext.js [hunk 5]
```diff
@@ -32,25 +32,31 @@ module.exports = class RequestContext {
...
-      relPath += 'index';
+    const lastSlash = this._path.lastIndexOf('/');
+    let lastDot = this._path.lastIndexOf('.');
+    if (lastDot <= lastSlash) {
+      // no extension means a folder request
+      const index = this._cfg.directoryIndex || 'index.html';
+      this._path = `${this._path}/${index}`;
+      // remove multipe slashes
+      this._path = this._path.replace(/\/+/g, '/');
+      lastDot = this._path.lastIndexOf('.');
```
Comment (tripodsan) [inline src/RequestContext.js]: you also need to adjust the lastSlash, since you need it further down for the selector calculation.

**Comment by tripodsan** [comment]
let's fix this for now, and then improve it later with #76

File: src/RequestContext.js (hunk 5 above)
Comment (tripodsan) [reply src/RequestContext.js]: but I think only if the `directoryIndex` would contain a selector. probably not important for now.

**Pull request merged**